packages/excubiae/contracts/extensions/HatsExcubia.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import {Excubia} from "../Excubia.sol";
+import {IHatsMinimal} from "./interfaces/IHatsMinimal.sol";
+
+/// @title Hats Excubia Contract.
+/// @notice This contract extends the Excubia contract to integrate with the Hats protocol.
+/// This contract checks if a user is wearing a specific hat to permit access through the gate.
+/// @dev The contract uses a specific set of hats to admit the passerby wearing any of those hats.
+contract HatsExcubia is Excubia {
+    /// @notice The Hats contract interface.
+    IHatsMinimal public immutable HATS;
+
+    /// @notice Mapping to track which hats are considered valid for passing the gate.
+    mapping(uint256 => bool) public criterionHat;
+    /// @notice Mapping to track which users have already passed through the gate.
+    mapping(address => bool) public passedUsers;
+
+    /// @notice Error thrown when the user is not wearing the required hat.
+    error NotWearingCriterionHat();
+    /// @notice Error thrown when the specified hat is not a criterion hat.
+    error NotCriterionHat();
+    /// @notice Error thrown when the array of criterion hats is empty.
+    error ZeroCriterionHats();
+
+    /// @notice Constructor to initialize the contract with the target Hats contract and criterion hats.
+    /// @param _hats The address of the Hats contract.
+    /// @param _criterionHats An array of hat IDs that are considered as criteria for passing the gate.
+    constructor(address _hats, uint256[] memory _criterionHats) {
+        if (_hats == address(0)) revert ZeroAddress();
+        if (_criterionHats.length == 0) revert ZeroCriterionHats();
+
+        HATS = IHatsMinimal(_hats);
+
+        for (uint256 i; i < _criterionHats.length; ++i) {
```

Comment:
should be cheaper to cache the length before the loop and pass that as counter end condition